### PR TITLE
network, broker [NET-518]: replace StreamIdAndPartition with SPID

### PR DESCRIPTION
**Pull request by teogeb**
Use `SPID` and `SPIDKey` to identify and a stream partition. Removed deprecated classes/types: `StreamdIdAndPartition`, `StreamKey` and `StreamPart`.

Most changes are modifications to internal data structures.  

External modifications: 

1. Status message format changed from `{ streamKey: '0x123::0, ...' }` to `{ id: '0x123', partition: 0 ... }`. There is a temporary backwards compatibility handler so that Tracker can read the deprecated format.
2. The data format of Tracker's HTTP endpoints changed. The keys are now `SPIDKeys`, not `StreamKeys`.
3. The public accessor method in in `Node.ts` and `Tracker.ts` changed from getStreams() to getSPIDs(). The return value changed from `string` array to Iterable<SPID>.

How SPID is used in this PR:

1. `SPID` is an independent entity, and a preferred way to reference to a stream partition. When we add a stream partition e.g. to a storage config, we use method `addSPID()` (not `getStreams()` or `getStreamPartitions()`). 
2. `SPIDKey` is an internal data format. The public methods of the application (in `Node.ts` and `Tracker.ts`) return `SPIDs`, not `SPIDKeys`. If we need to use textual representation of a SPID (e.g. as a JSON key, or as a parameter in command line), we read and write data in `SPIDKey` format.
3. We don't yet have public API methods which take `SPIDs` as input parameters. Maybe we could allow both: `SPID|SPIDKey`? 
4. In protocol level (the status message), we don't use `SPID` or `SPIDKey`. The format is plain `{ id, partition }`. 
5. In log files we don't use terms `SPID` or `SPIDKey`. There we use the term "partition": e.g.`"Add 123 partitions to storage config"`.

The internal hashing in `TrackerRegistry` are calculated from `SPIDKey` values. Therefore `streamId` -> `trackerIndex` results will be different after this change.

### CODE REVIEW

File: packages/broker/src/Stream.ts [hunk 1]
```diff
@@ -1,3 +1,6 @@
+import { Protocol } from 'streamr-network'
+import { SPID } from 'streamr-network/dist/streamr-protocol'
```
Comment (harbu) [inline packages/broker/src/Stream.ts]: Should import from `Protocol`?

File: packages/broker/src/StreamStateManager.ts [hunk 2]
```diff
@@ -37,7 +34,7 @@ export class StreamStateManager<C> {
             throw new Error('streamId or streamPartition not given!')
         }
 
-        const key = getStreamLookupKey(streamId, streamPartition)
+        const key = new Protocol.SPID(streamId, streamPartition).toKey()
```
Comment (harbu) [inline packages/broker/src/StreamStateManager.ts]: I wonder if we should make static method `SPID#toKey`   also support passing two arguments. So we can replace instances of 
```ts
const key = new Protocol.SPID(streamId, streamPartition).toKey()
```
with
```ts
const key = SPID.toKey(streamId, streamPartition)
```

It seems to be a common enough pattern to perhaps warrant doing?

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 3]
```diff
@@ -1,38 +1,19 @@
 import fetch from 'node-fetch'
-import { Logger } from 'streamr-network'
-import { StreamPart } from '../../types'
-import { StreamMessage, keyToArrayIndex } from 'streamr-network/dist/streamr-protocol'
+import { Logger, Protocol } from 'streamr-network'
+import { SPID } from 'streamr-network/dist/streamr-protocol'
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StorageConfig.ts]: Import

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 4]
```diff
@@ -1,38 +1,19 @@
...
-    return {
-        id,
-        partition: Number(partitionStr)
-    }
-}
-
-const getKeyFromStream = (streamId: string, streamPartition: number): StreamKey => {
-    return `${streamId}::${streamPartition}`
+    onSPIDAdded: (spid: Protocol.SPID) => void
+    onSPIDRemoved: (spid: Protocol.SPID) => void
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StorageConfig.ts:9]: I'm wonder if it would be more clear to refer to the stream partitions by full name?
```suggestion
    onStreamPartitionAdded: (spid: Protocol.SPID) => void
    onStreamPartitionRemoved: (spid: Protocol.SPID) => void
```

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 5]
```diff
@@ -49,7 +30,7 @@ export class StorageConfig {
 
     static ASSIGNMENT_EVENT_STREAM_ID_SUFFIX = '/storage-node-assignments'
 
-    streamKeys: Set<StreamKey>
+    private spidKeys: Set<Protocol.SPIDKey>
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StorageConfig.ts:32]: Again, wonder whether it would be more clear to use full name?
```suggestion
    private streamPartitions: Set<Protocol.SPIDKey>
```

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 6]
```diff
@@ -100,13 +85,13 @@ export class StorageConfig {
         this.poller = setTimeout(() => this.poll(pollInterval), pollInterval)
     }
 
-    hasStream(stream: StreamPart): boolean {
-        const key = getKeyFromStream(stream.id, stream.partition)
-        return this.streamKeys.has(key)
+    hasSPID(spid: Protocol.SPID): boolean {
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StorageConfig.ts:87]: Pondering whether full name would be clearer?
```suggestion
    hasStreamPartition(spid: Protocol.SPID): boolean {
```

File: packages/broker/src/broker.ts [hunk 7]
```diff
@@ -20,7 +19,7 @@ const logger = new Logger(module)
 
 export interface Broker {
     getNeighbors: () => readonly string[]
-    getStreams: () => readonly string[]
+    getSPIDs: () => Iterable<Protocol.SPID>
```
Comment (harbu) [inline packages/broker/src/broker.ts:22]: I wonder if we should refer to them as `StreamPartitions` in names?
```suggestion
    getStreamPartitions: () => Iterable<Protocol.SPID>
```

File: packages/broker/src/plugins/storage/StorageConfig.ts [hunk 8]
```diff
@@ -100,13 +85,13 @@ export class StorageConfig {
...
+        const key = new Protocol.SPID(spid.streamId, spid.streamPartition).toKey()
+        return this.spidKeys.has(key)
     }
 
-    getStreams(): StreamPart[] {
-        return Array.from(this.streamKeys).map((key) => getStreamFromKey(key))
+    getSPIDs(): Protocol.SPID[] {
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StorageConfig.ts:92]:
```suggestion
    getStreamPartitions(): Protocol.SPID[] {
```

File: packages/broker/src/plugins/storage/StoragePlugin.ts [hunk 9]
```diff
@@ -46,20 +45,16 @@ export class StoragePlugin extends Plugin<StoragePluginConfig> {
         this.cassandra = await this.getCassandraStorage()
         this.storageConfig = await this.createStorageConfig()
         this.messageListener = (msg) => {
-            const streamPart = {
-                id: msg.messageId.streamId,
-                partition: msg.messageId.streamPartition
-            }
-            if (this.storageConfig!.hasStream(streamPart)) {
+            if (this.storageConfig!.hasSPID(Protocol.SPID.from(msg.messageId))) {
```
Comment (harbu) [inline packages/broker/src/plugins/storage/StoragePlugin.ts]: Would it be useful to add a method to `StreamMessage` for retrieving the SPID? Seems like a use case that will come up a lot.

File: packages/broker/test/unit/plugins/storage/StorageConfig.test.ts [hunk 10]
```diff
@@ -1,6 +1,7 @@
 /* eslint-disable no-underscore-dangle, object-curly-newline */
 import { StorageConfig, StorageConfigListener } from '../../../../src/plugins/storage/StorageConfig'
 import nock from 'nock'
+import { SPID } from 'streamr-network/dist/streamr-protocol'
```
Comment (harbu) [inline packages/broker/test/unit/plugins/storage/StorageConfig.test.ts]: import

File: packages/network/src/helpers/transformIterable.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export const transformIterable = function *<F,T>(from: Iterable<F>, tranform: (f: F) => T): Iterable<T> {
```
Comment (harbu) [inline packages/network/src/helpers/transformIterable.ts:1]: Is this the equivalent of `Array#map` but of `Iterable`?

File: packages/network/src/logic/node/StreamManager.ts [hunk 12]
```diff
@@ -16,16 +17,16 @@ function keyForDetector({ publisherId, msgChainId }: MessageLayer.MessageID) {
 }
 
 export class StreamManager {
-    private readonly streams: Map<string, StreamState> = new Map<string, StreamState>() // streamKey => {}
+    private readonly streams: Map<SPIDKey,StreamState> = new Map<SPIDKey,StreamState>() // spidKey => {}
```
Comment (harbu) [inline packages/network/src/logic/node/StreamManager.ts]: Comment no longer necessary (remnant from pre-TS days, I belive)
```suggestion
    private readonly streams: Map<SPIDKey,StreamState> = new Map<SPIDKey,StreamState>()
```

File: packages/network/src/logic/node/TrackerConnector.ts [hunk 13]
```diff
@@ -12,12 +12,12 @@ enum ConnectionState {
     ERROR
 }
 
-type GetStreamsFn = () => Iterable<StreamIdAndPartition>
+type GetSPIDsFn = () => Iterable<SPID>
```
Comment (harbu) [inline packages/network/src/logic/node/TrackerConnector.ts:15]: Maybe full name?
```suggestion
type GetStreamPartitionsFn = () => Iterable<SPID>
```

File: packages/network/src/logic/node/TrackerConnector.ts [hunk 14]
```diff
@@ -26,22 +26,22 @@ export class TrackerConnector {
...
         this.connectToTracker = connectToTracker
         this.disconnectFromTracker = disconnectFromTracker
         this.trackerRegistry = trackerRegistry
         this.maintenanceInterval = maintenanceInterval
         this.connectionStates = new Map()
     }
 
-    onNewStream(streamId: StreamIdAndPartition): void {
-        const trackerInfo = this.trackerRegistry.getTracker(streamId.id, streamId.partition)
+    onNewStream(streamId: SPID): void {
```
Comment (harbu) [inline packages/network/src/logic/node/TrackerConnector.ts]: Re-name to spid or streamPartition?

File: packages/network/src/logic/node/TrackerConnector.ts [hunk 15]
```diff
@@ -26,22 +26,22 @@ export class TrackerConnector {
...
+        const trackerInfo = this.trackerRegistry.getTracker(streamId.streamId, streamId.streamPartition)
```
Comment (harbu) [inline packages/network/src/logic/node/TrackerConnector.ts]: Wondering if we should support `TrackerRegistry#getTracker(spid: SPID)` ?

File: packages/network/src/logic/node/TrackerManager.ts [hunk 16]
```diff
@@ -21,14 +21,14 @@ interface NodeDescriptor {
 }
 
 interface Subscriber {
-    subscribeToStreamIfHaveNotYet: (streamId: StreamIdAndPartition, sendStatus?: boolean) => void
+    subscribeToStreamIfHaveNotYet: (spid: SPID, sendStatus?: boolean) => void
```
Comment (harbu) [inline packages/network/src/logic/node/TrackerManager.ts:24]: I'm a bit undecided if we should use `streamPartition` as names for these or `spid`?
```suggestion
    subscribeToStreamIfHaveNotYet: (streamPartition: SPID, sendStatus?: boolean) => void
```

File: packages/network/src/logic/tracker/InstructionCounter.ts [hunk 17]
```diff
@@ -1,43 +1,45 @@
-import { Status, StreamKey } from '../../identifiers'
+import { SPID, SPIDKey } from 'streamr-client-protocol'
+import { Status } from '../../identifiers'
 import { NodeId } from '../node/Node'
 
 export const COUNTER_UNSUBSCRIBE = -1
 
-type Counters = Record<NodeId,Record<StreamKey,number>>
+type Counters = Record<NodeId,Record<SPIDKey,number>>
```
Comment (harbu) [inline packages/network/src/logic/tracker/InstructionCounter.ts]: I find it hard to parse these without spaces tbh:
```suggestion
type Counters = Record<NodeId, Record<SPIDKey, number>>
```

File: packages/network/test/unit/deprecated-tracker-status.test.ts [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        const actualStatus = updateNodeOnStream.mock.calls[0][1]
+        expect(actualStatus).toContainEntries([
+            ['id', STREAM_ID],
+            ['partition', STREAM_PARTITION],
+            ['neighbors', [TARGET_NODE_1, TARGET_NODE_2]],
+            ['counter', COUNTER],
+        ])
+        expect(formAndSendInstructions).toBeCalledTimes(1)
+        expect(formAndSendInstructions).toBeCalledWith(SOURCE_NODE, spidKey)
+    }
```
Comment (harbu) [inline packages/network/test/unit/deprecated-tracker-status.test.ts]: Instead of having to rely on all this mocking (and internals that are bound to change), couldn't you just use something from`trackerSummaryUtils.ts` to check the state of the tracker after a node with deprecated status message joins?

File: packages/protocol/src/utils/SPID.ts [hunk 19]
```diff
@@ -13,6 +13,8 @@ import ValidationError from '../errors/ValidationError'
 
 type RequiredKeys<T, Keys extends keyof T> = Omit<T, Keys> & Required<Pick<T, Keys>>
 
+export type SPIDKey = string
```
Comment (harbu) [inline packages/protocol/src/utils/SPID.ts:16]: With template literal types (https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) we can actually do the following
```suggestion
export type SPIDKey = `${string}#${number}`
```

**Review by harbu — COMMENTED** [review]
Nice work!

### Discussion Points
- <del>How to refer to SPIDs in variable / method names? E.g. `getSpids()` or `getStreamPartitions()`. I'm not convinced which one would be more readable in the end, both have their trade-offs.</del> We will use SPID.

### Suggestions
- Make SPIDKey into template literal type for a bit tighter type checking
- Add static method `SPID#toKey(streamId: string, streamPartition: number)` for ergonomics
- We should probably have method `StreamMessage#getSPID()` for convenience?
- SPID-argument support for `TrackerRegistry`, i.e., `TrackerRegistry#getTracker(spid: SPID)`?

### Fixes
- Some places are importing from `streamr-network/dist/streamr-protocol`
- I think I noticed some `streamId` named variables containing `SPID`. Not critical, but something I think I saw a couple times.

File: packages/network/test/unit/deprecated-tracker-status.test.ts (hunk 18 above)
Comment (teogeb) [reply packages/network/test/unit/deprecated-tracker-status.test.ts]: This backwards compatibility is a temporary feature, which we'll remove very soon. Then can easily remove this test, too, as all the backwards compatibility handling is tested here.

File: packages/network/src/helpers/transformIterable.ts (hunk 11 above)
Comment (teogeb) [reply packages/network/src/helpers/transformIterable.ts:1]: Yes, and doing the mapping lazily. One possible name for the utility function could be `mapIterable`.

File: packages/network/src/logic/node/TrackerConnector.ts (hunk 15 above)
Comment (teogeb) [reply packages/network/src/logic/node/TrackerConnector.ts]: Changed  `getTracker` to get the parameter as `SPID`. 

Also the internal hashing in TrackerRegistry are calculated from `SPIDKey` values. Therefore `streamId` -> `trackerIndex` results will be different after this change.

File: packages/broker/src/plugins/storage/StoragePlugin.ts (hunk 9 above)
Comment (teogeb) [reply packages/broker/src/plugins/storage/StoragePlugin.ts]: There was already that method, now using it in network and broker code. Added `InstructionMessage.getSPID()`

File: packages/protocol/src/utils/SPID.ts (hunk 19 above)
Comment (teogeb) [reply packages/protocol/src/utils/SPID.ts:16]: Seems quite complicated to add that restriction as we have *many* input formats to support. If we can simplify `SPID.fromDefaults()`, it should be possible to add the support for literal type.

If I remember correctly `SPID.fromDefaults()` was planned to be a temporary helper method, which helps us migrate all  input format variants to use `SPID`. Especially there was need for `SPID.fromDefaults(streamId)` to help us to collect all "partition 0 is the default, if not specified otherwise" logic to one place (and later migrate those usages one by one).

@timoxley Do I remember correctly? Can we therefore drop `SPID.fromDefaults()` after we the migration?

**Review by harbu — APPROVED** [review]
LGTM!

File: packages/protocol/src/utils/SPID.ts (hunk 19 above)
Comment (teogeb) [reply packages/protocol/src/utils/SPID.ts:16]: Created tickets:
- https://linear.app/streamr/issue/NET-566/client-use-spid
- https://linear.app/streamr/issue/NET-567/spidkey-as-template-literal-type
